control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+    }
+
+    // Collect the symbols for any header field lists which are passed to
+    // standard externs in the action body.
+    // XXX(seth): This is more hardcoded stuff.
+    using CallStatement = IR::MethodCallStatement;
+    forAllMatching<CallStatement>(action, [&](const CallStatement* callStatement) {
+      auto call = callStatement->methodCall;
+      auto instance = P4::MethodInstance::resolve(call, refMap, typeMap);
+      if (!instance->is<P4::ExternFunction>()) return;
+      auto externFunction = instance->to<P4::ExternFunction>();
+      if (externFunction->method->name == "clone" ||
+          externFunction->method->name == "clone3") {
+        symbols.add(PISymbolType::HEADER_FIELD_LIST, "fl");
+      } else if (externFunction->method->name == "resubmit") {
+        auto name = externalNameFromTypeArgs(call->typeArguments, refMap, "resubmit");
+        symbols.add(PISymbolType::HEADER_FIELD_LIST, name);
+      } else if (externFunction->method->name == "recirculate") {
+        auto name = externalNameFromTypeArgs(call->typeArguments, refMap, "recirculate");
+        symbols.add(PISymbolType::HEADER_FIELD_LIST, name);
+      }
+    });
+  }
+}
+
+static void serializeFieldListToPI(P4InfoSerializer& serializer,
+                                   TypeMap* typeMap,
+                                   const cstring& fieldListName,
+                                   const IR::Expression* fieldList) {
+  auto fieldListType = typeMap->getType(fieldList, true);
+  if (!fieldListType->is<IR::Type_Tuple>()) {
+    ::error("Expected header field list '%1%' to be of tuple type", fieldList);
+    return;
+  }
+
+  auto tuple = fieldListType->to<IR::Type_Tuple>();
+  std::vector<const HeaderField*> fields;
+  forAllTupleFields(tuple, typeMap, [&](const HeaderField* field) {
+    fields.push_back(field);
+  });
+
+  serializer.addHeaderFieldList(fieldListName, fields);
+}
+
+static void serializeControlToPI(P4InfoSerializer& serializer,
+                                 const IR::ControlBlock* controlBlock,
+                                 ReferenceMap* refMap,
+                                 TypeMap* typeMap) {
+  CHECK_NULL(controlBlock);
+  CHECK_NULL(refMap);
+  CHECK_NULL(typeMap);
+
+  auto control = controlBlock->container;
+  CHECK_NULL(control);
+
+  // XXX(seth): These are hardcoded in the BMV2 JSON serializer. Ideally we'd do
+  // something better, but this feature is slated for removal anyway.
+  if (control->name == "ingress" || control->name == "egress") {
+    serializer.addSynthesizedExitTable(control);
+  }
+
+  for (auto declaration : *control->controlLocals) {
+    // Serialize actions and, implicitly, their parameters.
+    if (!declaration->is<IR::P4Action>()) continue;
+    auto action = declaration->to<IR::P4Action>();
+    serializer.addAction(action);
+
+    // Serialize any header field lists which are passed to standard externs in
+    // the action body.
+    // XXX(seth): This is more hardcoded stuff.
```

Comment:
Indeed, this is architecture-dependent.
But maybe this could be done in a generic way - at least for extern functions. What should matter is just the function parameter names, not the argument values. In fact, this is just a function implemented by the control-plane called from the data plane. One should not care what the arguments are in the callee.